### PR TITLE
Fix ACTNUM manipulation with EQUALS

### PR DESCRIPTION
**Pull request by atgeirr**
The existing code struggles with using EQUALS to make cells active, that were originally made inactive in ACTNUM. When we later in the processing discover this, data for the re-activated cells is already thrown away. With this PR, the EclipseGrid will have the correct ACTNUM field from its construction time. The FieldProps class has been used from within EclipseGrid as a utility to do this, since FieldProps already had all the necessary functionality. It is still necessary to update the ACTNUM after pore volumes have been calculated, that part of the code has not been changed.

No unit tests yet, therefore a draft PR for now.

### CODE REVIEW

**Comment by atgeirr** [comment]
jenkins build this please

**Review by bska — COMMENTED** [review]
This looks good to me, at least as an initial solution to the problem.  I **suppose** a slightly different approach would be to add something like
```C++
static std::vector<int>
FieldPropsManager::computeInitiallyActiveCells(const Deck& deck, const GridDims& gdims);
```
One benefit of this is that we would not need to expose the low-level `FieldProps` class to the rest of the system.  By the way, if we **do** want to expose `FieldProps` to the rest of the system we also need to add the header file to those headers being installed.

As part of this alternative I also consider decoupling the [`Box`]() class from the grid structure through a couple of call-back functions, e.g.,
```C++
class Box
{
public:
    using IsActive = std::function<bool(const std::size_t globalIdx)>;
    using ActiveIdx = std::function<std::size_t(const std::size_t)>;

    explicit Box(const GridDims& gdims, IsActive isActive, ActiveIdx activeIdx);

    // ...
};
```
I'll continue looking into new test cases now.

**Comment by blattms** [comment]
Just my two cents (without having really looked into this in detail):
Is this change of design really needed? It seems a bit invasive. Why does not the normal workflow for changing via FieldProps::handle_operation` work here?

**Comment by atgeirr** [comment]
> Is this change of design really needed? It seems a bit invasive. Why does the normal workflow for changing via FieldProps::handle_operation` not work here?

Because we throw away data all the time when we process the properties, if the data are not in active cells (the EclipseGrid class is used for this), and the initial set of active cells (in EclipseGrid) is set from the last ACTNUM keyword in the deck which may have as inactive cells that are later re-activated. This simplistic treatment of initial ACTNUM also fails to deal with the situation where the last ACTNUM is incomplete, for example having two ACTNUM for two separate parts of the grid, each within a BOX block. The test we have for this currently succeeds only because it is not a corner point grid case, and the cartesian construction always starts ACTNUM as all-active. This could have been a general solution, but would likely cause a lot of extra memory consumption for corner point cases with low fraction of active cells, so it was decided to instead process ACTNUM completely to get a better starting point. (The set of active cells can be further shrunk by 0 pore volumes or net to gross, but never grown.)

**Comment by atgeirr** [comment]
I believe this is now ready for review/merge. I have added new tests, and in the process eliminated some corner cases that could lead to throwing.

**Comment by atgeirr** [comment]
jenkins build this please

File: tests/parser/FieldPropsTests.cpp [hunk 1]
```diff
@@ -134,6 +134,58 @@ PERMX
...
+)";
+
+    Deck deck = Parser{}.parseString(deck_string);
+    EclipseGrid grid{GridDims{3, 3, 1}};
+    FieldProps fp(deck, grid);
+    std::vector<int> expected_actnum = { 1, 1, 1,
+                                         1, 1, 1,
+                                         0, 0, 1  };
+    auto actnum = fp.actnumRaw();
+    BOOST_TEST(actnum == expected_actnum);
```
Comment (bska) [inline tests/parser/FieldPropsTests.cpp:169]: Any reason not to write this in terms of `BOOST_CHECK_EQUAL_COLLECTIONS`?  That way you get additional information regarding positions if elements differ
```C++
BOOST_CHECK_EQUAL_COLLECTIONS(actnum.begin(), actnum.end(), expected_actnum.begin(), expected_actnum.end());
```

Comment (atgeirr) [reply tests/parser/FieldPropsTests.cpp:169]: I will change it.

Comment (atgeirr) [reply tests/parser/FieldPropsTests.cpp]: Done

**Comment by bska** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks for the updates.  This looks good to me now and the build check passes.  I'll merge into master.
